sound/soc/sof/ipc4-loader.c

```diff
@@ -265,8 +265,86 @@ static int sof_ipc4_query_fw_configuration(struct snd_sof_dev *sdev)
 	return sof_ipc4_fw_parse_ext_man(sdev, plat_data->fw, 0, 0);
 }
 
+static int sof_ipc4_load_library(struct snd_sof_dev *sdev, struct snd_sof_widget *swidget)
+{
+	struct sof_ipc4_fw_data *ipc4_data = sdev->private;
+	struct snd_sof_pdata *plat_data = sdev->pdata;
+	const guid_t *widget_uuid = &swidget->uuid;
+	const char *fw_filename;
+	size_t fw_offset;
+	int ret, i;
+
+	/* nothing to do if widget UUID is not set */
+	if (!guid_is_null(&swidget->uuid))
+		return 0;
+
+	/* nothing to do if module library is already loaded */
+	for (i = 0; i < ipc4_data->num_fw_modules; i++)
+		if (guid_equal(&ipc4_data->module_uuids[i], widget_uuid))
+			return 0;
+
+	/*
+	 * The required 3rd party module libraries are dictated by the topology. So they should be
+	 * bundled in the same folder along with the topology binary
+	 */
+	fw_filename = kasprintf(GFP_KERNEL, "%s/%pUL.bin",
+				plat_data->tplg_filename_prefix, widget_uuid);
+	if (!fw_filename)
+		return -ENOMEM;
+
+	/* Index 0 is reserved for base firmware */
+	for (i = 1; i < ipc4_data->max_fw_libs; i++)
+		if (!ipc4_data->module_lib_info[i].id)
+			break;
+
+	if (i >= ipc4_data->max_fw_libs) {
+		dev_err(sdev->dev, "Library ID: %u exceeds max value %u\n",
+			i, ipc4_data->max_fw_libs);
+		ret = -EBUSY;
+		goto err;
+	}
+
+	ipc4_data->module_lib_info[i].id = i;
+	swidget->lib_info = &ipc4_data->module_lib_info[i];
+
+	ret = request_firmware(&ipc4_data->module_lib_info[i].fw, fw_filename, sdev->dev);
+	if (ret < 0) {
+		dev_err(sdev->dev, "Library file '%s' not found\n", fw_filename);
+		goto err;
+	}
+
+	fw_offset = sof_ipc4_fw_parse_ext_man(sdev, ipc4_data->module_lib_info[i].fw, i, 0);
+	if (fw_offset < 0)
+		goto release;
+
+	ipc4_data->module_lib_info[i].fw_offset = fw_offset;
+
+	if (sof_ops(sdev)->load_library)
+		ret = sof_ops(sdev)->load_library(sdev, &ipc4_data->module_lib_info[i]);
+	else
+		ret = -EINVAL;
```

Comment:
should we move the test earlier in the function to simplify, e.g. around line 277?

if (!sof_ops(sdev)->load_library)
    return -EINVAL;